diff --git a/src/parser.rs b/src/parser.rs
index 1654014..a15c215 100644
--- a/src/parser.rs
+++ b/src/parser.rs
@@ -107,8 +107,13 @@ impl<'a> Parser<'a> {
         StmtNode { kind: StmtKind::Block(stmts) }
     }
 
-    // expr-stmt = expr ";"
+    // expr-stmt = expr? ";"
     fn expr_stmt(&mut self) -> StmtNode {
+        if self.tok_is(";") {
+            self.advance();
+            return StmtNode { kind: StmtKind::Block(Vec::new()) }
+        }
+
         let expr = self.expr();
         self.skip(";");
         StmtNode { kind: StmtKind::Expr(expr) }
diff --git a/test.sh b/test.sh
index b3c7e27..887b4e1 100755
--- a/test.sh
+++ b/test.sh
@@ -64,6 +64,7 @@ assert 2 '{ 1; return 2; 3; }'
 assert 3 '{ 1; 2; return 3; }'
 
 assert 3 '{ {1; {2;} return 3;} }'
+assert 5 '{ ;;; return 5; }'
 
 rm -f tmp tmp.s
 
